File: 069.Sqrt(x)/sqrt.py
class Solution(object):
    def mySqrt(self, x):
        """
        :type x: int
        :rtype: int
        """
        low = 0
        high = x
        while low <= high:
            mid = (low + high) / 2
            square = mid * mid
            if square == x:
                return mid
            if square < x:
                low = mid + 1
            else:
                high = mid - 1

        return square <= x and mid or mid-1

if __name__ == '__main__':
    print Solution().mySqrt(1025)

File: 069.Sqrt(x)/Sqrt.java
public class Sqrt {
    // 要用长整型避免溢出
    public int mySqrt(int x) {
        long low = 0, high = x;
        long square = 0, mid = 0;
        while (low <= high) {
            mid = (low + high) / 2;
            square = mid * mid;
            if (square == x) {
                return (int)mid;
            }
            if (square > x) {
                high = mid - 1;
            }
            else {
                low = mid + 1;
            }
        }

        return (int)(square <= x ? mid : mid-1);
    }

    public static void main(String[] args) {
        System.out.println(new Sqrt().mySqrt(2147395599));
    }
}
